### PR TITLE
<feat> custom head를 가진 reader model 코드 추가 + masking logic

### PR DESCRIPTION
**Pull request by olenmg**
## 요약
효과가 크게 없어서 지금까지 PR을 안날렸는데, 계속 commit 충돌이 나서 먼저 PR 해둡니다.
PR이전에 기존 DPR모델도 이전과 동일하게 돌아가는 것 확인했습니다!

계속 말씀드렸던 custom head 모델 코드입니다.
- LSTM/CNN/FC head를 추가하였습니다.
- 학습중에 토큰 masking을 할 수 있도록 추가하였습니다.
- **config 사용 형식이 바뀌었습니다. README 참고해주세요.**
- **(5/16추가)** 
    - training 중 evaluation 할 수 있도록 기능 추가하였습니다.
    - Complex CNN head 추가하였습니다. 

## 세부 설명
1. custom head를 추가하였습니다.
    - LSTM/CNN 모델은 임의로 추가한 모델입니다. 구조는 코드가 쉽게 쓰여있어 보시면 바로 아실 수 있습니다.
    - FC 모델은 head 모양 자체는 일반 DPR 모델과 동일합니다. ~다만 원래 모델은 768 x 2 linear layer 1개를 사용했었는데, 여기서는 클래스 상속을 위해 768 x 1 linear layer 2개 사용하고 있습니다. 별 문제는 없을 것 같습니다. 쓸지도 모르겠구요, 기존 모델과 비교를 위해 만들어두었습니다.~ **head에서 output 차원이 2가 되도록 코드 수정하였습니다.**
2. custom head를 사용할 수 있도록 MRC reader 모델 코드를 추가하였습니다.
    - 현재 custom model의 경우 checkpoint를 바로 불러오기 어려운 상태라, model_path 인자가 추가되었는데 custom model 사용 안하시면 신경 안쓰셔도 됩니다!
    - custom model 사용하실경우 model은 model 이름 주시고, model_path에 checkpoint의 pytorch_model.bin 경로 주시면 됩니다.
3. masking을 추가하였습니다.
    - SpanBERT에서 영감을 받아, 연속된 2개 토큰을 마스킹하도록 세팅되어있으나, 추후 어떻게 해볼지 아직 고민중입니다.
    - args에서 masking ratio를 줄 수 있습니다. custom에서만 동작하고, **DPR모델에서는 동작하지 않습니다.**
4. prepare.py에서 custom model 작동을 위해 일부 내용 수정하였으나, 마찬가지로 DPR 모델 돌리는데에는 지장 없습니다.
5. strategy에 인자 몇 개를 추가하였습니다. README 참고해주세요! 다만 custom model 돌리실거 아니면 신경 안쓰셔도 됩니다.
6. (추가) pororo reader부분도 커밋되었는데, git에서 인식을 못해서 commit한거고 제가 뭐 수정한건 아니라서 이전과 코드 완전히 동일합니다.
7. **(5/16추가)** 
    - training 중 evaluation 할 수 있도록 기능 추가하였습니다.
        - README에 업데이트해뒀는데, `do_eval_during_training` 옵션을 `true`로 주시고 validation을 하고 싶은 step을 `eval_step`으로 주시면 매 eval_step마다 `trainer.evaluate()`을 실행합니다.

## 그 외
**수정 예정**
1. predict.py에서 model_name_or_path로 바로 run/predict에서 checkpoint 불러올 수 있도록 수정

제가 근 1~2주간 Reader 모델만 맡았음에도 성능 향상이 거의 없어 너무 아쉽고 또 정말 죄송스럽습니다
다음주에는 정말 집중해서 성능 최대한 끌어올려보겠습니다

### CODE REVIEW

**Comment by SeongIkKim** [comment]
주말 새벽까지 PR하시느라 너무 고생많으셨습니다. 지금 종헌님이 맡고계신 파트가 너무 핵심적이고 어려운 파트라고 생각이 들고, 종헌님이 코드 수정하시고 추가하시는것도 너무 많으신데 죄송하게 생각할 필요는 전혀 없다고 생각합니다! 그보다는 제가 reader를 같이 보기로 했고 봐야하는데 주말간 다른 일정과 학교 과제에 치여서 같이 고민해보지 못해서 오히려 너무 죄송합니다. ㅠㅠ

좋은 성과를 일찍 내면 물론 좋겠지만, 아직 4주차가 남았으니 그 때까지 할 수 있는 시도들을 계속 해내가는 것도 좋지 않을까요?

PR 부분은 확인했고, 수정된 파일 대부분이 기존 코드에는 영향을 크게 주지 않을것같아서 다른 분들만 확인되시면 일단 merge 후에 월요일 간단하게 리뷰하면 될것같습니다!

**Comment by olenmg** [comment]
성익님 좋은 말씀 감사합니다ㅋㅋㅋㅠㅠ 사실 저 스스로 아쉬움이 많이 남네요 그래도 아직 일주일 남아있으니 열심히 해볼게요! 그리고 성익님께서도 후처리로 성능에 큰 기여하고 계신데 죄송하실 필요 전혀 없습니다 ㅋㅋㅋㅋ

사실 생각난게 있어서 잠시 코멘트달러 왔습니다🤣 fine-tuned 모델 안쓰면 오버피팅이 좀 덜할 것 같아서, 내일은 CNN모델의 복잡도를 조금 올려보겠습니다! 현재는 인풋 차원을 768->1로 너무 급작스럽게 줄이고 있는데 확실히 좋은 방법은 아닌 것 같네요. 내일은 좀더 보편화된 CNN 구조로 생각해보려고 합니다~ 

CNN을 고집하는것도 좋은 방법이 아닐 수 있지만, 제 개인적인 생각으론 지금 task를 start logit, end logit 뽑아서 풀 수 있는 최선의 방법은 CNN에서 나올 것 같아요.. 그게 아니면 아예 start, end 뽑는 형식이 아니라 아웃풋 자체가 다른 방법을 생각해봐야 할 것 같습니다. 근데 아웃풋이 다른 방법은 떠오르지 않네요

**Comment by ggm1207** [comment]
확인했습니다! 종헌님 지금 충분히 잘 하고 계십니다! 죄송해 할 것 전혀 없습니다!! 저도 성능 측면으로는 한 거 하나도 없지만 뻔뻔하게 잘 하고 있으니 .. :cry:

**Comment by sooyounlee** [comment]
앗 댓글 단줄알았는데 안달았었네요 ㅋㅋ 저도 확인했습니다!

**Comment by ebbunnim** [comment]
종헌님, 확인했습니다. 저야말로 오픈소스로 dense retriever 새로 생성하는거 못하고 있는데요 ㅠㅠ 제가 가장 죄송한 것 같습니다. 코드 굉장히 깔끔하게 작성하셔서 commit 내역 이해하기 쉬웠습니다. 감사합니다!
